### PR TITLE
fix(statesync): if statesync fails, handshake with the app

### LINKED ISSUE #4977: statesync: if statesync fails or reaches max_discovery_time, it must handshake with the app
Otherwise, the node would use the genesis state, which would be incorrect.

discovered in https://github.com/cometbft/cometbft/actions/runs/13426543714/job/37510433419

Introduced in https://github.com/cometbft/cometbft/pull/3965

### PR DESCRIPTION
**Pull request by melekes**
before starting blocksync.

Closes #4977

### CODE REVIEW

**Comment by melekes** [comment]
> Looks like there are a number of CodeQL messages here regarding logging the error object. Do these sometimes contain sensitive info?

no

File: node/node.go [hunk 1]
```diff
@@ -636,17 +646,66 @@ func (n *Node) OnStart() error {
...
+		stateCh, errc, err := startStateSync(n.stateSyncReactor, n.stateSyncProvider,
+			n.config.StateSync, n.stateStore, n.blockStore, n.stateSyncState, n.config.Storage.ExperimentalKeyLayout)
 		if err != nil {
 			return ErrStartStateSync{Err: err}
 		}
+
+		// Wait until statesync is over or errors, and start blocksync in a separate goroutine.
+		go func() {
+			defer func() {
+				// FREE GENESIS DOC MEMORY!
```
Comment (melekes) [inline node/node.go:665]: this line, I'm not proud, but don't think there's another way.

File: node/node.go [hunk 2]
```diff
@@ -636,17 +646,66 @@ func (n *Node) OnStart() error {
...
-			n.config.StateSync, n.stateStore, n.blockStore, n.stateSyncGenesis, n.config.Storage.ExperimentalKeyLayout)
+
+		stateCh, errc, err := startStateSync(n.stateSyncReactor, n.stateSyncProvider,
+			n.config.StateSync, n.stateStore, n.blockStore, n.stateSyncState, n.config.Storage.ExperimentalKeyLayout)
 		if err != nil {
 			return ErrStartStateSync{Err: err}
 		}
+
+		// Wait until statesync is over or errors, and start blocksync in a separate goroutine.
+		go func() {
```
Comment (jmalicevic) [inline node/node.go:663]: I was a bit unsure whether the parallelism here and the channels open a way for introducing errors but I can't think of a scenario atm.
The things I was trying to understand:
1. Can we have `state` returned on the `stateCh` before the Syncing function sets the syncing flag to false (which should not happen after we switch to blocksync) - but I think it is not possible. 
2. Could the note somehow setup wrongly, but again I am not seeing it with this code. 

Writing this just in case something clicks on your end when reading this :D

File: test/e2e/runner/setup.go [hunk 3]
```diff
@@ -324,7 +324,6 @@ func MakeConfig(node *e2e.Node) (*config.Config, error) {
 		if len(cfg.StateSync.RPCServers) < 2 {
 			return nil, errors.New("unable to find 2 suitable state sync RPC servers")
 		}
-		cfg.StateSync.MaxDiscoveryTime = 30 * time.Second
```
Comment (jmalicevic) [inline test/e2e/runner/setup.go:327]: Are we sure that by removing this we are testing all possible corner cases of statesync now?

File: node/node.go [hunk 4]
```diff
@@ -636,17 +646,66 @@ func (n *Node) OnStart() error {
...
+			case newState := <-stateCh:
+				n.Logger.Info("statesync complete. switching to blocksync with new state")
+
+				err = bcR.SwitchToBlockSync(newState)
+				if err != nil {
+					n.Logger.Error("failed to switch to blocksync", "err", err)
+					return
+				}
+			case err := <-errc:
+				n.Logger.Warn("statesync failed. switching to blocksync with genesis state", "err", err)
```
[the remark this answers is not in the document]
Comment (melekes) [reply node/node.go:682]: Don't see why we can't try blocksync if statesync fails. We log the error. I.e. logging the error and trying blocksync is better than ONLY logging the error (note Comet doesn't exit upon statesync error).

File: test/e2e/runner/setup.go (hunk 3 above)
Comment (melekes) [reply test/e2e/runner/setup.go:327]: Do you know why it was introduced in the first place?

File: node/node.go (hunk 2 above)
Comment (melekes) [reply node/node.go:663]: I will change it to wait. I just noticed that we're sleeping in `Node.Start` 😄

Comment (jmalicevic) [reply node/node.go]: Where is the sleep?

**Review by zrbecker — APPROVED** [review]

File: node/node.go (hunk 2 above)
Comment (jmalicevic) [reply node/node.go]: But is this not only an exception and an error case while in the regular case tehre is no sleep?

Comment (melekes) [reply node/node.go]: True. Still avoiding extra goroutines gives app developers more control over the `Start` function.

Comment (jmalicevic) [reply node/node.go]: Tbh I would avoid changing the behaviour that was there for some years w/o being sure it hsa no implications. For example, someone expecting `node.Start()` to return w/o statesync might be doing other things that could facilitate StateSync execution. 

So I would put statesync in a separate go routine as it was.
